### ISSUE TITLE
xctool will not run under fastlane

### ISSUE THREAD

**Issue by bmctigue** [fastlane/fastlane#4170]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)

### Issue Description

xctool will not run under fastlane. Fastlane is updated to latest gems and xctool is also updated. xctool runs fine locally external to fastlane running the command generated by fastlane.

##### Complete output when running fastlane, including the stack trace and command used

```
Bruces-MacBook-Pro:betterup-ios bmctigue$ bin/deploy-lint
[08:20:30]: Driving the lane 'lint' 🚀
[08:20:30]: -----------------------------------
[08:20:30]: --- Step: clean_build_artifacts ---
[08:20:30]: -----------------------------------
[08:20:30]: Cleaned up build artifacts 🐙
[08:20:30]: -------------------------------------
[08:20:30]: --- Step: ensure_git_status_clean ---
[08:20:30]: -------------------------------------
[08:20:30]: Git status is clean, all good! 💪
[08:20:30]: --------------------
[08:20:30]: --- Step: xctool ---
[08:20:30]: --------------------
[08:20:30]: Have you seen the new 'scan' tool to run tests? https://github.com/fastlane/fastlane/tree/master/scan
[08:20:30]: $ xctool --workspace 'betterup.xcworkspace' --scheme 'betterup-testflight' --reporter json-compilation-database:compile_commands.json clean build
[08:20:44]: Variable Dump:
[08:20:44]: {:PLATFORM_NAME=>nil, :LANE_NAME=>"lint", :GIT_REPO_WAS_CLEAN_ON_START=>true}
[08:20:44]: Exit status of command 'xctool --workspace 'betterup.xcworkspace' --scheme 'betterup-testflight' --reporter json-compilation-database:compile_commands.json clean build' was 1 instead of 0.


+------+-------------------------+-------------+
|               fastlane summary               |
+------+-------------------------+-------------+
| Step | Action                  | Time (in s) |
+------+-------------------------+-------------+
| 1    | clean_build_artifacts   | 0           |
| 2    | ensure_git_status_clean | 0           |
| 3    | xctool                  | 14          |
+------+-------------------------+-------------+

[08:20:44]: fastlane finished with errors
Successfully submitted a crash report. If this is a problem with one of the tools specifically,
please submit an issue on GitHub and attach the following number to it: 'cce4b2ae61fa40fa125d0df1f3e7502c'
The crash report has been stored locally '/tmp/sentry_cce4b2ae61fa40fa125d0df1f3e7502c.json'
/Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane_core-0.41.2/lib/fastlane_core/ui/interface.rb:127:in `crash!': [!] Exit status of command 'xctool --workspace 'betterup.xcworkspace' --scheme 'betterup-testflight' --reporter json-compilation-database:compile_commands.json clean build' was 1 instead of 0. (FastlaneCore::Interface::FastlaneCrash)

	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane_core-0.41.2/lib/fastlane_core/ui/ui.rb:14:in `method_missing'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/helper/sh_helper.rb:44:in `sh_no_action'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/helper/sh_helper.rb:9:in `sh'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/actions/xctool.rb:12:in `run'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/runner.rb:143:in `block (2 levels) in execute_action'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/actions/actions_helper.rb:35:in `execute_action'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/runner.rb:129:in `block in execute_action'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/runner.rb:128:in `chdir'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/runner.rb:128:in `execute_action'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/fast_file.rb:149:in `method_missing'
	from Fastfile:44:in `run_oclint'
	from Fastfile:83:in `block in parsing_binding'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/lane.rb:36:in `call'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/lane.rb:36:in `call'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/runner.rb:52:in `block in execute'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/runner.rb:46:in `chdir'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/runner.rb:46:in `execute'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/lane_manager.rb:46:in `cruise_lane'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/lib/fastlane/command_line_handler.rb:30:in `handle'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/bin/fastlane:38:in `block (2 levels) in run'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/commander-4.3.5/lib/commander/command.rb:178:in `call'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/commander-4.3.5/lib/commander/command.rb:178:in `call'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/commander-4.3.5/lib/commander/command.rb:153:in `run'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/commander-4.3.5/lib/commander/runner.rb:428:in `run_active_command'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane_core-0.41.2/lib/fastlane_core/ui/fastlane_runner.rb:23:in `run!'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/commander-4.3.5/lib/commander/delegates.rb:15:in `run!'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/bin/fastlane:168:in `run'
	from /Users/bmctigue/.rvm/rubies/ruby-2.2.1/lib/ruby/gems/2.2.0/gems/fastlane-1.81.0/bin/fastlane:174:in `<top (required)>'
	from bin/fastlane:16:in `load'
	from bin/fastlane:16:in `<main>'
Bruces-MacBook-Pro:betterup-ios bmctigue$ 
```

### Configuration Files

Please copy the complete content of your `Fastfile` and any other configuration files you use below:

Fastfile:

```ruby
# Customise this file, documentation can be found here:
# https://github.com/krausefx/fastlane#customise-the-fastfile

# All lines starting with a # are ignored when running `fastlane`
# You can define as many lanes as you want

before_all do
  clean_build_artifacts
  ensure_git_status_clean
end

def before_package
  scan(scheme:"betterup-testflight", skip_slack:true, device:"iPhone 6")
  cert
  sigh
end

def before_internal
  scan(scheme:"betterup-testflight", skip_slack:true, device:"iPhone 6")
  increment_build_number
  cert
  sigh
end

def before_release
  scan(scheme: "betterup", skip_slack:true, device:"iPhone 6")
  increment_build_number
  cert
  sigh
end

def after_package
  clean_build_artifacts
end

def after_release
  clean_build_artifacts
  commit_version_bump
  add_git_tag
  push_to_git_remote
end

def run_oclint
  xctool [
      "--workspace", "'betterup.xcworkspace'",
      "--scheme", "'betterup-testflight'",
      "--reporter", "json-compilation-database:compile_commands.json",
      "clean","build"
  ].join(" ")
  oclint(
    compile_commands: 'compile_commands.json',
    report_path: 'oclint_report.html',
    select_regex: /.*.m/,
    exclude_regex: /(Pods|betterup\/vendor|betterup\/categories|betterup\/camera|betterup\/vendor|betterup\/betterupTests)/,
    disable_rules: [
      'BitwiseOperatorInConditional',
      'ConstantIfExpression',
      'UnusedMethodParameter',
      'TooManyMethods'
    ],
    max_priority_1: 10,                   # The max allowed number of priority 1 violations
    max_priority_2: 1000,                  # The max allowed number of priority 2 violations
    max_priority_3: 1000,                 # The max allowed number of priority 3 violations
    thresholds: [
      'LONG_VARIABLE_NAME=30',
      'LONG_LINE=270',
      'LONG_METHOD=50'
    ],
    allow_duplicated_violations: false
  )
end

def coverage
  xcov(
    workspace: "betterup.xcworkspace",
    scheme: "betterup-testflight",
    output_directory: "xcov_output",
    skip_slack: true
  )
end

lane :lint do
  run_oclint
end

lane :coverage do
  coverage
end

lane :test do
  scan(scheme:"betterup-testflight", skip_slack:true, device:"iPhone 6")
  # we need to uncomment this as soon as circle can run xctool
  # run_oclint
  coverage
end

lane :preview do
  snapshot
  frameit
end

lane :package do
  before_package
  # verify_xcode

  gym(
    workspace: "betterup.xcworkspace",
    configuration: "Release",
    scheme: "betterup",
    output_name: 'betterup.ipa',
    codesigning_identity: ENV['SIGN_IDENTITY']
  )
  
  after_package
end

lane :internal do
  before_internal
  verify_xcode

  gym(
    workspace: "betterup.xcworkspace",
    configuration: "Release Testflight",
    scheme: "betterup-testflight",
    output_name: 'betterup.ipa',
    codesigning_identity: ENV['SIGN_IDENTITY']
  )

  pilot

  slack(
    message: "New internal build #{Actions.lane_context[Actions::SharedValues::BUILD_NUMBER]} available from Testflight",
    success: true,
    channel: 'development'
  )
  after_release
end

lane :public do
  before_release
  verify_xcode

  gym(
    workspace: "betterup.xcworkspace",
    configuration: "Release",
    scheme: "betterup",
    output_name: 'betterup.ipa',
    codesigning_identity: ENV['SIGN_IDENTITY']
  )

  deliver

  slack(
    message: "New public build #{Actions.lane_context[Actions::SharedValues::BUILD_NUMBER]} uploaded to iTunesConnect",
    success: true,
    channel: 'development'
  )
  increment_version_number(
    bump_type: "patch" # Automatically increment patch version number
  )
  after_release
end

# This block is called, only if the executed lane was successful
after_all do |lane|
end

# Something bad happened
error do |lane, exception|
end
```

fastlane version (fastlane -v): fastlane 1.81.0
Do you use bundler, rbenv or ram? bundler

**Comment by vpolouchkine** [comment]
I think you need to pass an xctool action. Take a look at https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Actions.md#xctool

**Comment by bmctigue** [comment]
Thanks for the response. This used to work until recently without any action. Is there an action for ocLint? I'm not using xctool for testing in this case. Again, xctool runs from the command line just fine with no action.

**Comment by TKBurner** [comment]
Hi @bmctigue! Thanks for following up. Can you run xcode-select -v and send me the output?

Thanks!

**Comment by bmctigue** [comment]
Here you go...

xcode-select version 2343.

**Comment by TKBurner** [comment]
@bmctigue Thanks for following up. We're going to look into this over here. I'm going to mark this as help-wanted in case anyone from the community has advice!

**Comment by mfurtak** [comment]
Hi @bmctigue - Thanks for the report :+1: 

The `xctool` call you're making in your `Fastfile` is calling through to the fastlane `xctool` action, but that should all be the same as it ever was.

Because the `xctool` action is an incredibly thin wrapper around shelling-out to the `xctool` command line tool, there's really not a whole lot that can go wrong on the fastlane side. The error message you're seeing is a result of the generated shell-out command returning a non-zero exit status, which we interpret to mean that it failed at doing what you requested it to do.

According to `xctool`, exit code 1 means ["XCToolActionFailed"](https://github.com/facebook/xctool/blob/master/xctool/xctool/XCTool.h#L21)

I know you mentioned that things are working when you run it outside of fastlane, but if you could run this exact generated command from your logas (plus the diagnostic bit at the end), it would really help!

I'd be interested to know whether running:

```
xctool --workspace 'betterup.xcworkspace' --scheme 'betterup-testflight' --reporter json-compilation-database:compile_commands.json clean build && echo 'Success!'
```

prints 'Success!' at the end of the run or not. If not, `xctool` is not completing successfully.

Thanks!

**Comment by KrauseFx** [comment]
Does that work for you @bmctigue?

**Comment by bmctigue** [comment]
Hi, 
Thanks for the follow up. That command did not work for me. I did find a fix for me though. This command is now working (generated by fastlane) xctool --workspace 'betterup.xcworkspace' --scheme 'betterup-testflight' --reporter json-compilation-database:compile_commands.json clean build. I had to apply this obscure fix to Xcode to get this to run:
https://forums.developer.apple.com/thread/25248. Not sure if anyone else is having this issue, but it does fix my issue with xctool. I suppose we can close this issue for me, but I have no idea why this all of a sudden was necessary to apply. Maybe you guys have more insight. Thanks for looking into this.
